GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,195 @@
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Text;
+using System.Threading;
+
+namespace GVFS.Common.Tracing
+{
+    public enum QueuedPipeStringWriterState
+    {
+        Unknown = 0,
+        Stopped = 1,
+        Failing = 2,
+        Healthy = 3,
+    }
+
+    /// <summary>
+    /// Accepts string messages from multiple threads and dispatches them over a named pipe from a
+    /// background thread.
+    /// </summary>
+    public class QueuedPipeStringWriter : IDisposable
+    {
+        private const int DEFAULT_MAX_QUEUE_SIZE = 256;
+
+        private readonly Func<NamedPipeClientStream> createPipeFunc;
+        private readonly BlockingCollection<string> queue;
+
+        private IQueuedPipeStringWriterEventSink eventSink;
+        private Thread writerThread;
+        private NamedPipeClientStream pipeClient;
+        private QueuedPipeStringWriterState state = QueuedPipeStringWriterState.Unknown;
+        private bool isDisposed;
+
+        public QueuedPipeStringWriter(Func<NamedPipeClientStream> createPipeFunc, int maxQueueSize = DEFAULT_MAX_QUEUE_SIZE)
+        {
+            this.createPipeFunc = createPipeFunc;
+            this.queue = new BlockingCollection<string>(new ConcurrentQueue<string>(), boundedCapacity: maxQueueSize);
+        }
+
+        public void RegisterEventSink(IQueuedPipeStringWriterEventSink eventSink)
```

Comment:
Again can this interface come in the constructor so we don't need this method or the check for already being registered?

Reply:
Done!